### PR TITLE
Feature: Structured logging support (JSON)

### PR DESCRIPTION
**Pull request by SadiHassan**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

This feature will provide JSON formatted log output in console. The purpose is to allow easier ingestion and query in log analysis tools.

1. A sample log output might look like this.
```{"datetime":"2022.03.28 12:17:54.827035","thread_id":"790151","query_id":"2441139","priority":"Debug","source":"KeeperTCPHandler","message":"Received close event with xid 2147483647 for session id #7"}```
2. User will be able to turn on/off `JSON` output.
For example consider the sample tag below under ```<logger>``` (that will be written in `config.xml`). 
```
<format>
     <json>true</json>
</format>
```
If user deletes or comments out the `format` or `json` tag, usual log output will prevail. 

Note: This is the first stage of implementation of #35366. As @blinkov suggested, once this part gets accepted, the other part can be implemented right after.
...

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by SadiHassan** [comment]
@nikitamikhaylov I see stylecheck failed for the PR description template modified by me. Did you add "Can be tested" for your later manual test? (I mean, manually run other tests that checks coding and functionality?)

File: base/loggers/Loggers.cpp [hunk 1]
```diff
@@ -49,7 +50,6 @@ void Loggers::buildLoggers(Poco::Util::AbstractConfiguration & config, Poco::Log
         if (auto log = text_log.lock())
             split->addTextLog(log, text_log_max_priority);
 #endif
-
```
Comment (alexey-milovidov) [inline base/loggers/Loggers.cpp:52]: This change looks irrelevant.

File: base/loggers/Loggers.h [hunk 2]
```diff
@@ -10,15 +10,17 @@
...
 }
 #endif
 
 namespace Poco::Util
 {
-    class AbstractConfiguration;
+class AbstractConfiguration;
 }
 
+extern bool log_format_json;
```
Comment (alexey-milovidov) [inline base/loggers/Loggers.h:22]: Global variables makes code feel bad.

File: base/loggers/OwnPatternFormatter.cpp [hunk 3]
```diff
@@ -7,13 +7,118 @@
...
+    writeCString(",", wb);
+
+    writeCString("\"", wb);
+    writeCString("thread_id", wb);
+    writeCString("\"", wb);
+    writeCString(":", wb);
+    writeCString("\"", wb);
+    writeCString(" [ ", wb);
+    if (color)
+        writeString(setColor(intHash64(msg_ext.thread_id)), wb);
```
Comment (alexey-milovidov) [inline base/loggers/OwnPatternFormatter.cpp:48]: JSON does not have colors.

File: base/loggers/OwnPatternFormatter.cpp [hunk 4]
```diff
@@ -7,13 +7,118 @@
...
+    DB::writeChar('0' + ((msg_ext.time_microseconds / 10000) % 10), wb);
+    DB::writeChar('0' + ((msg_ext.time_microseconds / 1000) % 10), wb);
+    DB::writeChar('0' + ((msg_ext.time_microseconds / 100) % 10), wb);
+    DB::writeChar('0' + ((msg_ext.time_microseconds / 10) % 10), wb);
+    DB::writeChar('0' + ((msg_ext.time_microseconds / 1) % 10), wb);
+    writeCString("\"", wb);
+
+    writeCString(",", wb);
+
+    writeCString("\"", wb);
```
Comment (alexey-milovidov) [inline base/loggers/OwnPatternFormatter.cpp:41]: writeChar can be used.

File: base/loggers/OwnPatternFormatter.cpp [hunk 5]
```diff
@@ -7,13 +7,118 @@
...
+    writeCString("thread_id", wb);
+    writeCString("\"", wb);
+    writeCString(":", wb);
+    writeCString("\"", wb);
+    writeCString(" [ ", wb);
```
Comment (alexey-milovidov) [inline base/loggers/OwnPatternFormatter.cpp:46]: There is no sense to put decorated string ` [1]` instead of number.

File: base/loggers/OwnPatternFormatter.cpp [hunk 6]
```diff
@@ -7,13 +7,118 @@
...
+    writeCString("\"", wb);
+    writeCString("query_id", wb);
+    writeCString("\"", wb);
+    writeCString(":", wb);
+    writeCString("\"", wb);
+    writeCString("{", wb);
+
+    if (color)
+        writeString(setColor(std::hash<std::string>()(msg_ext.query_id)), wb);
+    DB::writeString(msg_ext.query_id, wb);
```
Comment (alexey-milovidov) [inline base/loggers/OwnPatternFormatter.cpp:68]: This is wrong, because the string is not escaped.

File: base/loggers/OwnPatternFormatter.cpp [hunk 7]
```diff
@@ -7,13 +7,118 @@
...
+    writeCString("\"", wb);
+    writeCString("source", wb);
+    writeCString("\"", wb);
+    writeCString(":", wb);
+
+    writeCString("\"", wb);
+
+    if (color)
+        writeString(setColor(std::hash<std::string>()(msg.getSource())), wb);
+    DB::writeString(msg.getSource(), wb);
```
Comment (alexey-milovidov) [inline base/loggers/OwnPatternFormatter.cpp:104]: The string is not escaped.

File: base/loggers/OwnPatternFormatter.cpp [hunk 8]
```diff
@@ -7,13 +7,118 @@
...
+
+    writeCString(",", wb);
+
+    writeCString("\"", wb);
+    writeCString("message", wb);
+    writeCString("\"", wb);
+    writeCString(":", wb);
+
+    writeCString("\"", wb);
+    DB::writeString(msg.getText(), wb);
```
Comment (alexey-milovidov) [inline base/loggers/OwnPatternFormatter.cpp:117]: The string is not escaped.

File: base/loggers/OwnPatternFormatter.cpp [hunk 9]
```diff
@@ -7,13 +7,118 @@
...
+    writeCString("\"", wb);
+
+    writeCString("}", wb);
```
Comment (alexey-milovidov) [inline base/loggers/OwnPatternFormatter.cpp:120]: Missing file, line and other fields from system.text_log.

File: programs/server/config.xml [hunk 10]
```diff
@@ -60,6 +60,18 @@
           </logger>
         </levels>
         -->
+        <!-- Structured log formatting:
+        
+        You can specify log format (for now, JSON only). In that case, the console log will be printed
+        in JSON format.
+        For example, like this below:
+        {"datetime":"2022.03.28 12:17:54.827035","thread_id":"790151","query_id":"2441139","priority":"Debug","source":"KeeperTCPHandler","message":"Received close event with xid 2147483647 for session id #7"}
```
Comment (alexey-milovidov) [inline programs/server/config.xml:68]: DateTime should be in unix timestamp or in ISO-8601 with UTC.

Comment (alexey-milovidov) [inline programs/server/config.xml:68]: query_id does not typically looks like this.

File: programs/server/config.xml [hunk 11]
```diff
@@ -60,6 +60,18 @@
...
+        You can specify log format (for now, JSON only). In that case, the console log will be printed
+        in JSON format.
+        For example, like this below:
+        {"datetime":"2022.03.28 12:17:54.827035","thread_id":"790151","query_id":"2441139","priority":"Debug","source":"KeeperTCPHandler","message":"Received close event with xid 2147483647 for session id #7"}  
+        To enable JSON logging support, just uncomment <format> tag and add <json> tag. 
+        Having the <json> tag will make it work. For better understanding/visibility, you can add "true" or "1".
+        -->
+        <format>
+            <json>true</json>
+        </format>
```
Comment (alexey-milovidov) [inline programs/server/config.xml:74]: Need to write a test.

Comment (alexey-milovidov) [reply programs/server/config.xml:74]: For better consistency, let's make JSON formatting exactly as JSONEachRow formatting of system.text_log.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: base/loggers/Loggers.cpp (hunk 1 above)
Comment (SadiHassan) [reply base/loggers/Loggers.cpp:52]: Agree. Will remove.

File: base/loggers/Loggers.h (hunk 2 above)
Comment (SadiHassan) [reply base/loggers/Loggers.h:22]: The problem here is, we need to get the config settings which is found in `Loggers.cpp`. We also need to use the `config` properties into `OwnFormattingChannel` and `OwnPatternFormatter`. But there is no data passing from `Loggers` to any of them that passes the `config` property. That's why global was used. However, I found another way that is adding a static member inside `ExtendedLogMessage` class and accessing that member from `Loggers.cpp`. This is slightly better than global namespace. This will keep it under `DB::ExtendedLogMessage`. Do you have any suggestion around this?

File: base/loggers/OwnPatternFormatter.cpp (hunk 3 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp:48]: Agreed. Will change it.

File: base/loggers/OwnPatternFormatter.cpp (hunk 4 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp:41]: Agreed. Will change this.

File: base/loggers/OwnPatternFormatter.cpp (hunk 5 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp:46]: Agreed. Will change this. When I found test case failing because of `assert` in python that checks specific format with `[` or `{`, I added those to make test pass.

File: base/loggers/OwnPatternFormatter.cpp (hunk 6 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp:68]: Agreed. Will change this.

File: base/loggers/OwnPatternFormatter.cpp (hunk 7 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp:104]: Agreed. Will change this.

File: base/loggers/OwnPatternFormatter.cpp (hunk 8 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp:117]: Agreed. Will change this.

File: base/loggers/OwnPatternFormatter.cpp (hunk 9 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp:120]: Agreed. Will add them. Is there any suggestion of their sequence?

File: programs/server/config.xml (hunk 10 above)
Comment (SadiHassan) [reply programs/server/config.xml:68]: Agreed. Will change this.

Comment (SadiHassan) [reply programs/server/config.xml:68]: Agreed. This was a made up thing to fill the explanation. Will update with originals.

File: programs/server/config.xml (hunk 11 above)
Comment (SadiHassan) [reply programs/server/config.xml:74]: Any idea of what kind of test should it run? My guess is, if the `config` has `json` in it, the test program should grab log message and see if the message is in correct log format or not. Anything else? Also, can you hint any test file to understand how to write a test function for this?

File: base/loggers/OwnPatternFormatter.cpp (hunk 9 above)
Comment (alexey-milovidov) [reply base/loggers/OwnPatternFormatter.cpp:120]: The ordering can be identical to the `system.text_log` table.

File: programs/server/config.xml (hunk 11 above)
Comment (alexey-milovidov) [reply programs/server/config.xml:74]: I can be integration test in `tests/integration`.
It can check that server writes logs in JSON, that JSON is valid, and not necessarily anything else.

**Comment by alexey-milovidov** [comment]
Please address review comments.

**Comment by alexey-milovidov** [comment]
Abandoned.

File: base/loggers/Loggers.h (hunk 2 above)
Comment (SadiHassan) [reply base/loggers/Loggers.h:22]: change submitted in updated PR.

File: base/loggers/OwnPatternFormatter.cpp (hunk 3 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp]: change submitted in updated PR.

File: base/loggers/OwnPatternFormatter.cpp (hunk 4 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp]: change submitted in updated PR.

File: base/loggers/OwnPatternFormatter.cpp (hunk 5 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp]: change submitted in updated PR.

File: base/loggers/OwnPatternFormatter.cpp (hunk 6 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp]: change submitted in updated PR.

File: base/loggers/OwnPatternFormatter.cpp (hunk 7 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp]: change submitted in updated PR.

File: base/loggers/OwnPatternFormatter.cpp (hunk 8 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp]: change submitted in updated PR.

File: base/loggers/OwnPatternFormatter.cpp (hunk 9 above)
Comment (SadiHassan) [reply base/loggers/OwnPatternFormatter.cpp]: change submitted in updated PR.

File: programs/server/config.xml (hunk 10 above)
Comment (SadiHassan) [reply programs/server/config.xml]: change submitted in updated PR.

Comment (SadiHassan) [reply programs/server/config.xml]: change submitted in updated PR.

File: base/loggers/OwnFormattingChannel.cpp [hunk 12]
```diff
@@ -1,18 +1,23 @@
 #include "OwnFormattingChannel.h"
+#include <iostream> //mallik--remove pls
```
Comment (larryluogit) [inline base/loggers/OwnFormattingChannel.cpp:2]: ^^

**Comment by SadiHassan** [comment]
Hi @alexey-milovidov, I suspect the failing test cases are not related to the changes I made. Can you please have a look and suggest?

**Comment by SadiHassan** [comment]
@alexey-milovidov submitted the enhanced version of this feature in another PR #36896. Once this one is reviewed, you can take a look at that.

**Comment by alexey-milovidov** [comment]
@SadiHassan 

Instead of using global variable, we can implement it in a different `PatternFormatter`, similar to `OwnPatternFormatter` (you can name it `OwnJSONPatternFormatter`) and create it in `BaseDaemon.cpp` according to configuration.

**Comment by SadiHassan** [comment]
> @SadiHassan
> 
> Instead of using global variable, we can implement it in a different `PatternFormatter`, similar to `OwnPatternFormatter` (you can name it `OwnJSONPatternFormatter`) and create it in `BaseDaemon.cpp` according to configuration.
@alexey-milovidov 
Okay. In that case, we will have to make changes in the following way: According to configuration, we will have to add `OwnJSONPatternFormatter` in `BaseDaemon.cpp` (1 place, inside `setupWatchdog()` function) and in `Loggers.cpp` (3 places, under conditions of `logpath`, `errorlog_path` and `use_syslog`. Also, to make it work, we need and overloaded constructor of `OwnFormattingChannel` which will take `OwnJSONPatternFormatter` as a parameter (instead of `OwnPatternFormatter`). Finally, in `OwnFormattingChannel.cpp`, we have to call our `formatExtendedJSON()` function based on `pFormatterJSON` is not null. 
Please share your thoughts about this.

**Comment by SadiHassan** [comment]
@alexey-milovidov as discussed, removed global variable and created new class `OwnJSONPatternFormatter` and applied necessary changes in files as explained above.
